Automattic/wp-calypso
# Checkout: Improve types of checkout pending and post-checkout redirect

#### Proposed Changes

This PR makes relatively small changes to the `CheckoutPending` component (the "payment pending" page that's sometimes displayed while a transaction processes) and the `getThankYouPageUrl()` function (used by checkout to determine the post-checkout redirect URL) to clarify the types that they use. There should not be any functional changes, although some modifications to the code were necessary so there's a risk. I'll enumerate the code changes inline.

Besides adding additional protection against regressions, this also may help us eventually use the `CheckoutPending` for _every_ purchase as a way to track completed purchases which we otherwise can't easily do on the front-end (as described in https://github.com/Automattic/wp-calypso/issues/53356).

#### Testing Instructions

The significant changes should already be covered by unit tests (and this PR also converts some of those unit tests to TypeScript to further that safety).

One thing to check is that the redirect payment processors still work as expected since those often use the "pending" feature that this PR modifies. Here's instructions for Bancontact:

- Apply D45443-code to force Bancontact to be available and sandbox the API.
- Change your user's currency to `EUR`.
- Add a product to your cart and visit checkout.
- Complete the checkout form and select "Bancontact" as the payment method.
- Fill out the "Name" field and submit the payment. You should be redirected to a Stripe test page.
- Click to accept the payment. You should be redirected to a "Processing..." page briefly (this is the pending component) before being redirected to some post-checkout page (which page will vary based on many factors but is not that important).

The pending page:

<img width="1512" alt="Screen Shot 2022-07-02 at 7 36 01 PM" src="https://user-images.githubusercontent.com/2036909/177019068-dfde9949-eeef-47f4-b707-1f3aa54b5526.png">
